Add readme and swagger specification

ianbrown9475 opened the pull request.

Commit:
Add readme and swagger specification

Commit:
Fixed schemas in swagger.yaml

Commit:
Further changes to swagger.yaml

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,298 @@
+      - "application/json"
       parameters:
-        - $ref: '#/parameters/pretty'
+        - name: "id"
```
Why have an `id` query parameter if you also have a /cards/`id` endpoint? You don't have to change anything, just curious

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,298 @@
+          description: "Invalid request"
+          schema:
+            $ref: "#/definitions/Error"
+        404:
```
I'm not sure if this endpoint should return a 404. In jsonAPI spec, endpoints that usually return an array of objects would return a 200 with an empty `"data"` array if no cards were found that matched the query. http://jsonapi.org/format/#fetching-resources-responses-200

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,298 @@
+      tags:
+      - "cards"
+      summary: "Retrieve cards based on search parameters"
+      description: ""
```
Looks like you aren't using the description fields for any of the endpoints. Could you remove them instead?

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,298 @@
+            $ref: "#/definitions/SingleData"
+
+      responses:
+        201:
```
For POST and PUT, it would be a good idea to return the object that was added or updated in the http response.

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,298 @@
+          schema:
+            $ref: "#/definitions/SingleData"
+      responses:
+        204:
```
Same comment as the POST, it would be good to return the updated object in the response. I also made a mistake with my comment in Jira when recommending you change the status code. I realized a 200 might be the better option because there isn't a good code for "updated".

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,298 @@
+      responses:
+        204:
+          description: "Card successfully removed"
+        400:
```
What type of request might return a `400` for a `DELETE`?

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,298 @@
+
+        - name: "keywords"
+          in: "query"
+          description: "List of keywords found in a card's description"
```
Just curious, what will this query parameter be searching? I didn't see a description field for the card object.

ianbrown9475 committed:
Fix swagger.yaml according to comments

ianbrown9475 commented:
- id parameter removed (redundant due to /cards/{id} endpoint

ianbrown9475 commented:
- Removed id parameter in GET /cards due to redundancy with /cards/{id} endpoint
- GET /cards no longer returns a 404 due to no cards matching
- Removed empty description fields
- PUT, POST, and DELETE now respond with an HTTP 200 and return the SingleData object that was updated, added, or removed respectively
- DELETE no longer returns an HTTP 400 because there is no syntax in the body that can be incorrect
- Added a "description" field in the Card object
- Added an isRandom parameter (boolean) in the GET /cards endpoint that is meant to return cards in a random order if true. This is meant to emulate drawing cards of a certain type, color, ...
- Added NewData object used by the POST /cards endpoint that lacks an id field for adding a new card. Structure is otherwise equivalent to SingleData
- Added an HTTP 500 response to all endpoints in the event of an internal server error

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,330 @@
+          description: "Card object to be added to the database"
+          required: true
+          schema:
+            $ref: "#/definitions/SingleData"
```
Should this be using the `NewData` object too, similar to put?

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,330 @@
+            $ref: "#/definitions/SingleData"
+
       responses:
         200:
```
Perhaps a `201` would be more appropriate when creating a new card object? Sorry if my comment didn't make sense originally.

jaredkosanovic commented on swagger.yaml:
```diff
@@ -1,75 +1,330 @@
+        properties:
+          type:
+            type: "string"
+          attributes:
```
The attributes for NewData and Card are the same. Maybe use one shared object instead?

jaredkosanovic commented on swagger.yaml:
```diff
@@ -240,14 +244,16 @@ paths:
           format: "int64"
       responses:
-        204:
+        200:
```
Just curious, why did you change this from a 204? I would say a 204 is more common when deleting a resource, but either are acceptable.

ianbrown9475 committed:
swagger.yaml: restructured objects, minor response changes

ianbrown9475 commented:
- POST /cards now uses NewData in body instead of PUT /cards/{id}
- PUT /cards/{id} uses SingleData in compliance with jsonapi: http://jsonapi.org/format/#crud-updating
- POST /cards responds with a 201 rather than a 200 for resource creation
- Restructured objects to factor properties better
- Successful DELETE responds with a 204 and no content rather than responding with a 200 and including the deleted object
